dataclasses.py

```diff
@@ -528,13 +529,34 @@ class C:
 
       c = C(1, 2)
       assert asdict(c) == {'x': 1, 'y': 2}
+
+    If given, 'dict_factory' will be used instead of built-in dict.
+    The function applies recursively to field values that are
+    dataclass instances. This will also look into built-in containers: tuples,
+    lists, and dicts.
     """
-    if isinstance(obj, type):
-        raise ValueError("asdict() should be called on dataclass instances, not classes")
-    return {name: getattr(obj, name) for name in fields(obj)}
+    if isinstance(obj, type) or not hasattr(obj, _MARKER):
+        raise TypeError("asdict() should be called on dataclass instances")
+    return _asdict_inner(obj, dict_factory)
+
+def _asdict_inner(obj, dict_factory):
+    result = []
+    for name in fields(obj):
+        value = getattr(obj, name)
+        if hasattr(value, _MARKER) and not isinstance(value, type):
+            value = _asdict_inner(value, dict_factory)
+        elif isinstance(value, (list, tuple)):
+            value = type(value)(_asdict_inner(v, dict_factory) for v in value)
```

Comment:
You need to make sure that v is a dataclass here before recursing (either that, or have a guard at the top of this function. Consider:
```
@dataclass
class Child:
    d: object

@dataclass
class Parent:
    child: Child

asdict(Parent(Child([1])))
```
Gives:
```
Traceback (most recent call last):
  File "f.py", line 13, in <module>
    asdict(Parent(Child([1])))
  File "/cygdrive/c/home/eric/local/dataclasses/dataclasses.py", line 540, in asdict
    return _asdict_inner(obj, dict_factory)
  File "/cygdrive/c/home/eric/local/dataclasses/dataclasses.py", line 547, in _asdict_inner
    value = _asdict_inner(value, dict_factory)
  File "/cygdrive/c/home/eric/local/dataclasses/dataclasses.py", line 549, in _asdict_inner
    value = type(value)(_asdict_inner(v, dict_factory) for v in value)
  File "/cygdrive/c/home/eric/local/dataclasses/dataclasses.py", line 549, in <genexpr>
    value = type(value)(_asdict_inner(v, dict_factory) for v in value)
  File "/cygdrive/c/home/eric/local/dataclasses/dataclasses.py", line 544, in _asdict_inner
    for name in fields(obj):
  File "/cygdrive/c/home/eric/local/dataclasses/dataclasses.py", line 518, in fields
    return getattr(cls, _MARKER)
AttributeError: 'int' object has no attribute '__dataclass_fields__'
```